codemp/rd-rend2/tr_image.cpp

```diff
@@ -2569,6 +2581,155 @@ void R_UpdateSubImage( image_t *image, byte *pic, int x, int y, int width, int h
 		ri.Hunk_FreeTempMemory( resampledBuffer );
 }
 
+#define GAMMA 2.2f
+#define INV_GAMMA 1.0 / GAMMA
+
+void R_CreateDiffuseAndSpecMapsFromBaseColorAndRMO(shaderStage_t *stage, const char *name, const char *rmoName, int flags, int type)
+{
+	image_t	*image;
+	char	diffuseName[MAX_QPATH];
+	char	specularName[MAX_QPATH];
+	int		width, height, rmoWidth, rmoHeight;
+	byte	*rmoPic, *baseColorPic, *specGlossPic, *diffusePic;
+	long	hash;
+
+	if (!name) {
+		return;
+	}
+
+	COM_StripExtension(name, diffuseName, MAX_QPATH);
+	Q_strcat(diffuseName, MAX_QPATH, "_diffuse");
+
+	COM_StripExtension(name, specularName, MAX_QPATH);
+	Q_strcat(specularName, MAX_QPATH, "_spec");
+
+	//
+	// see if the images are already loaded
+	//
+	hash = generateHashValue(diffuseName);
+	for (image = hashTable[hash]; image; image = image->next) {
+		if (!strcmp(diffuseName, image->imgName)) {
+			stage->bundle[TB_COLORMAP].image[0] = image;
+			// check for specular map
+			hash = generateHashValue(specularName);
+			for (image = hashTable[hash]; image; image = image->next) {
+				if (!strcmp(specularName, image->imgName)) {
+					stage->bundle[TB_SPECULARMAP].image[0] = image;
+					return;
+				}
+			}
+		}
+	}
+
+	//
+	// load the pics from disk
+	//
+	R_LoadImage(name, &baseColorPic, &width, &height);
```

Comment:
You're leaking memory at each of these returns (you need to free the images that were loaded before)